### ISSUE TITLE
Type inference incorrect when it depends on a Higher Ranked Trait Bound impl

### ISSUE THREAD

**Issue by daboross** [intellij-rust/intellij-rust#4176]
## Environment

* **IntelliJ Rust plugin version:**
* **Rust toolchain version:**
* **IDE name and version:**
* **Operating system:**

## Problem description

I've run into an issue that's fairly roundabout, but I believe is caused by something to due with HRTB.

![screenshot-20190723-180248](https://user-images.githubusercontent.com/1152146/61757124-16181780-ad74-11e9-82f7-417e252bb7e9.png)

As shown, using `.clone()` here infers using the implementation of `impl<T> Clone for &T`, not the implementation for `X`.

If I remove the higher-ranked bound `for<'de> serde::Deserialize<'de>`, replacing it with a non-HRTB bound, then `b` is correctly inferred to be type `X`.

In my codebase, this mainly causes false-positive "incorrect type" errors when passing cloned values around. It seems it's willing to believe me if I use `X::clone` explicitly, but I'd like to not need to do that (and the real type names are much longer).

## Steps to reproduce

Create a new Rust project, and replace the main/lib file with the following code:

```rust
struct X;
impl Clone for X where String: for<'de> serde::Deserialize<'de> {
    fn clone(&self) -> Self { X }
}

fn test() {
    let a = &X;
    let b = a.clone();
}
```

Enable parameter type hints, so that the type of `b` is shown. `b` is shown as being of type `&X`, rather than `X`.

Alternative, include the following instead:

```rust
struct X;

impl Clone for X where String: for<'de> serde::Deserialize<'de> {
    fn clone(&self) -> Self { X }
}

fn func(x: X) {}

fn test() {
    let a = &X;
    func(a.clone());
}
```

`func(a.clone());` is then incorrectly shown to be a type mismatch.

**Comment by daboross** [comment]
Apologies, I seem to have misplaced the actual error. I'm super hoping this isn't just "IntelliJ can't parse some feature `serde` uses to implement `Serialize` / `Deserialize`", so if it is that, I apologize.

**Comment by daboross** [comment]
My actual use case for this is a structure closer to this:

```
pub struct Handle<T: Message>(PhantomData<T>);

impl<T: Message> Clone for Handle<T> { ... }

pub trait Message: Serialize + for<'de> Deserialize<'de> + fmt::Debug {}

impl<T> Message for T where T: Serialize + for<'de> Deserialize<'de> + fmt::Debug {}
```

I could get rid of the bound on struct and thus also on the `Clone` impl, so this isn't too high priority for me. But ideally I'd like to keep both.

---

If solving "tell if a type is `serde::Serialize`" is a hard problem, maybe a more practical solution would be assuming `T` does implement `Message` if `T` appears in `Handle<T>`?

Or assuming `impl<T: Message> Clone for Handle<T>` is always-application if `Handle` also has the `T: Message` bound.
